internal/sql_runner.go

```diff
@@ -38,36 +43,141 @@ var errorConnectionStates = []string{
 }
 
-// SQLRunner is a runner for run the sql.
-type SQLRunner struct {
-	db *sql.DB
+var internalLog = log.Log.WithName("mysql-internal")
+
+// Config is used to connect to a MysqlCluster.
+type Config struct {
+	User     string
+	Password string
+	Host     string
+	Port     int32
+}
+
+// NewConfigFromClusterKey returns a new Config based on a MySQLCluster key.
+func NewConfigFromClusterKey(c client.Client, clusterKey client.ObjectKey, userName, host string) (*Config, error) {
+	cluster := &apiv1alpha1.Cluster{}
+	if err := c.Get(context.TODO(), clusterKey, cluster); err != nil {
+		return nil, err
+	}
+
+	secret := &corev1.Secret{}
+	secretKey := client.ObjectKey{Name: mysqlcluster.New(cluster).GetNameForResource(utils.Secret), Namespace: cluster.Namespace}
+
+	if err := c.Get(context.TODO(), secretKey, secret); err != nil {
+		return nil, err
+	}
+
+	if host == utils.LeaderHost {
+		host = fmt.Sprintf("%s-leader.%s", cluster.Name, cluster.Namespace)
+	}
+
+	switch userName {
+	case utils.OperatorUser:
+		password, ok := secret.Data["operator-password"]
+		if !ok {
+			return nil, fmt.Errorf("operator-password cannot be empty")
+		}
+		return &Config{
+			User:     utils.OperatorUser,
+			Password: string(password),
+			Host:     host,
+			Port:     utils.MysqlPort,
+		}, nil
+
+		
+	case utils.RootUser:
+		password, ok := secret.Data["root-password"]
+		if !ok {
+			return nil, fmt.Errorf("root-password cannot be empty")
+		}
+		return &Config{
+			User:     utils.RootUser,
+			Password: string(password),
+			Host:     host,
+			Port:     utils.MysqlPort,
+		}, nil
+	default:
+		return nil, fmt.Errorf("MySQL user %s are not supported", userName)
+	}
+
 }
 
-// NewSQLRunner return a pointer to SQLRunner.
-func NewSQLRunner(user, password, host string, port int) (*SQLRunner, error) {
-	dataSourceName := fmt.Sprintf("%s:%s@tcp(%s:%d)/?timeout=5s&interpolateParams=true&multiStatements=true",
-		user, password, host, port,
+// GetMysqlDSN returns a data source name.
+func (c *Config) GetMysqlDSN() string {
+	return fmt.Sprintf("%s:%s@tcp(%s:%d)/?timeout=5s&multiStatements=true&interpolateParams=true",
+		c.User, c.Password, c.Host, c.Port,
 	)
-	db, err := sql.Open("mysql", dataSourceName)
+}
+
+type sqlRunner struct {
+	db *sql.DB
+}
+
+// SQLRunner interface is a subset of mysql.DB.
+type SQLRunner interface {
+	QueryExec(query Query) error
+	QueryRow(query Query, dest ...interface{}) error
+	QueryRows(query Query) (*sql.Rows, error)
+}
+
+// SQLRunnerFactory a function that generates a new SQLRunner.
+type SQLRunnerFactory func(cfg *Config, errs ...error) (SQLRunner, func(), error)
+
+// NewSQLRunner opens a connections using the given DSN.
+func NewSQLRunner(cfg *Config, errs ...error) (SQLRunner, func(), error) {
+	var db *sql.DB
+	var close func()
+
+	// Make this factory accept a functions that tries to generate a config.
+	if len(errs) > 0 && errs[0] != nil {
+		return nil, close, errs[0]
```

Comment:
return close is correct ？

Reply:
In this position, close and nil are equivalent, I think close can be used to return.